### PR TITLE
Resort projects when creating or updating their name

### PR DESCRIPTION
**Pull request by oliverguenther**
- [x] Reorders a saved project when creating, or when their name is being changed.
- [x] Add a reorder job that reorders every child project
- [x] Adds a migration to call that job once
- [x] Adds a sort order for typeahead which uses lft and name

https://community.openproject.org/work_packages/40930

### CODE REVIEW

File: app/workers/projects/reorder_children_job.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    end
+
+    def unique_parent_ids
+      Project
+        .where.not(parent_id: nil)
+        .select(:parent_id)
+        .distinct
+    end
+
+    def reorder_children(parent)
```
Comment (machisuji) [inline app/workers/projects/reorder_children_job.rb:53]: Wouldn't a `Project.rebuild!` have the same effect since we have configured the sorting column in `acts_as_nested_set`?

**Review by machisuji — APPROVED** [review]
LGTM 👍

File: db/migrate/20220202140507_reorder_project_children.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class ReorderProjectChildren < ActiveRecord::Migration[6.1]
+  def up
+    ::Projects::ReorderChildrenJob.perform_later
```
Comment (wielinde) [inline db/migrate/20220202140507_reorder_project_children.rb:3]: @oliverguenther Using `perform_later` will show only an effect when the job workers are running. That isn't always the case, especially when debugging it. It might create confusion. I'd argue it is safe to use `perfom_now`. Only then the modification of the data is actually done during the migration for sure.

**Review by wielinde — APPROVED** [review]
@oliverguenther Looks good. I checked the new order on Postman. Also ran checked the migration. There I ran into the `perform_later` vs. `perform_now` issue that I've commented. You might want to change that.

You already commented to me that there still is some Ruby based sorting in use when ordering  the projects for the project selector. Would be great if we could get rid of that piece of code if it is not needed anymore.

File: db/migrate/20220202140507_reorder_project_children.rb (hunk 2 above)
Comment (oliverguenther) [reply db/migrate/20220202140507_reorder_project_children.rb:3]: Thanks, I was deliberately using perform_later here to not increase the migration time on SaaS. And for on-premises installations, the worker will need to run anyway so there will only be a slight delay to the sorting. But I'm fine to make the migrations run longer for most customers except SaaS. @machisuji What do you think?

Comment (wielinde) [reply db/migrate/20220202140507_reorder_project_children.rb:3]: I get the point of speeding up the migration. I don't believe that it takes long per tenant but I cannot say to how much it sums up when migrating whole shards. My discomfort comes with the unclear state the data is in. I always considered migrations to leave the DB at a well defined state. And here it is not as the DB's state depends on when the job workers run. Let's say you run a on boarding tour right after the migration that trusts that the first project is the demo project but it is not there yet. Then the on boarding brakes unnecessarily... But yes, it is hypothetical.

Comment (machisuji) [reply db/migrate/20220202140507_reorder_project_children.rb:3]: I always dislike any code being run at all during database migrations. We've had plenty of cases where code which was used in an old migration had disappeared by the time a customer was migrated (e.g. from OP7 to 11). Using a job mitigates this somewhat at least as we only rely on the presence of the job and not a specific method signature.
And should the job not be needed anymore you could just make it a no-op.

Of course this still is all true when you use `.perform_now`. 
But since this job doesn't change the database structure itself it doesn't make a difference for following migrations if it's run (right away) or not.
  I agree with Wieland that the database should be in a consistent state after a migration. But in this case it is with or without the reordering. We're not changing the structure of the database here. This is a 'data migration'.
  Also running it later does speed up the migration on SaaS. So it makes sense to me.

The projects are still there and in the right hierarchy too. It's just the order within each level that'll change once the job has run and all this is touching is the lft and rgt attributes of the projects. So in my book it's perfectly fine to do this later in a job.

tl;dr; I'd prefer the job in the background

Comment (oliverguenther) [reply db/migrate/20220202140507_reorder_project_children.rb:3]: Thanks for your feedback. I'll merge this for now and suggest we revisit this topic with @ulferts when he's back
